### PR TITLE
[JENKINS-62757] Introduce fingerprint migration in External Storage API

### PR DESCRIPTION
**Pull request by stellargo**
See [JENKINS-62757](https://issues.jenkins-ci.org/browse/JENKINS-62757).
See [JEP-226](https://github.com/jenkinsci/jep/tree/master/jep/226).

### Proposed changelog entries

* Developer: Migration of fingerprints from local file based storage to configured external storage is introduced.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by stellargo** [comment]
This PR is WIP currently.

**Comment by stellargo** [comment]
@oleg-nenashev @afalko @mikecirioli This is the migration PR. Work is done but suggest putting it on hold till https://github.com/jenkinsci/jenkins/pull/4834 is merged.

File: test/src/test/java/jenkins/fingerprints/FingerprintStorageTest.java
Comment (afalko) [inline test/src/test/java/jenkins/fingerprints/FingerprintStorageTest.java:118]: Shouldn't this be private? I'm curious why not have a mock for this object.

File: core/src/main/java/hudson/model/Fingerprint.java
Comment (afalko) [inline core/src/main/java/hudson/model/Fingerprint.java:1320]:
```suggestion
     * In case an external storage is configured on top of a file system based storage: 

     1) External storage is...
     2) If not found and fingerprint is present...
```

This description is difficult to read, see above suggestion on how to refactor it (basically don't have one long run-on sentence here :))

**Review by afalko — CHANGES_REQUESTED** [review]
Small question and comment improvement suggestion

File: test/src/test/java/jenkins/fingerprints/FingerprintStorageTest.java
Comment (stellargo) [reply test/src/test/java/jenkins/fingerprints/FingerprintStorageTest.java:118]: I want to re-use this class in another test also. Would you suggest creating this in a separate file? I think that would be better.
So would you suggest using something like Mockito for this?

Comment (afalko) [reply test/src/test/java/jenkins/fingerprints/FingerprintStorageTest.java:118]: I think using Mockito might be better (you get coverage on the actual class instead of your test class) and lighter-weight, no?

Comment (stellargo) [reply test/src/test/java/jenkins/fingerprints/FingerprintStorageTest.java:118]: The only problem I think is that I want to test the functionality too. IIUC Mockito will give me a placeholder, but it cannot act as an external storage, which I want to test if it is working or not.

**Comment by stellargo** [comment]
@afalko @oleg-nenashev @mikecirioli request for review :)

**Review by timja — APPROVED** [review]

File: core/src/main/java/hudson/model/Fingerprint.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Fingerprint.java:1330]: Would be better to cache the storage implementations as local variables (current storage and then the Filesystem one). Apart from performance benefits, it reduces the chance of the logic exploding in unlikely case when the Storage gets reconfigured in the middle of the `load()` operation. Multi-threading 🤷

File: core/src/main/java/hudson/model/Fingerprint.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Fingerprint.java:1380]: Same as above

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I would recommend to use local variables to make the code more reliable against concurrency issues

File: core/src/main/java/hudson/model/Fingerprint.java
Comment (stellargo) [reply core/src/main/java/hudson/model/Fingerprint.java]: I agree, thanks for pointing it out :) Done!

**Review by oleg-nenashev — APPROVED** [review]
Should be good to go, thanks @stellargo !

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Review by oleg-nenashev — APPROVED** [review]
👍 for catching the bug @stellargo . Resetting the merge timer

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Pull request merged**
